### ISSUE TITLE
teacher model in eval() mode but still update gradients?

### ISSUE THREAD

**Issue by yuezha01** [peterliht/knowledge-distillation-pytorch#25]
Hi,

Very useful code and instructions! If I understand it correctly, the teacher model shouldn't be updated with gradients and only the student model will compute gradients during the distillation process. I noticed in the `train_and_evaluate_kd()` function, the teacher model is set to eval() mode. But I think eval() only alters the behavior of dropout or BatchNorm, it doesn't stop gradient update when loss.backward() is called. I think teacher model's parameters should set require_grad to False.

**Comment by yuezha01** [comment]
[Here](https://github.com/peterliht/knowledge-distillation-pytorch/blob/e4c40132fed5a45e39a6ef7a77b15e5d389186f8/train.py#L164) the output from the teacher model is saved as numpy arrays. Then the arrays are converted to tensors to compute the loss so I think the teacher model is not involved.
